model/lifecycle.go

```diff
@@ -320,39 +320,50 @@ func RestartTasksInVersion(ctx context.Context, versionId string, abortInProgres
 // RestartVersion restarts completed tasks associated with a versionId.
 // If abortInProgress is true, it also sets the abort flag on any in-progress tasks.
 func RestartVersion(ctx context.Context, versionId string, taskIds []string, abortInProgress bool, caller string) error {
+	// Fetch tasks to reset before aborting to ensure we're including any needed display tasks.
+	tasks, err := getTasksWithDisplayTasks(taskIds)
+	if err != nil {
+		return errors.Wrap(err, "getting finished tasks")
+	}
+	tasksToReset := []task.Task{}
+	abortIds := []string{}
+	for i, t := range tasks {
+		if utility.StringSliceContains(evergreen.TaskCompletedStatuses, t.Status) {
+			tasksToReset = append(tasksToReset, tasks[i])
+		} else {
+			abortIds = append(abortIds, t.Id)
+		}
+	}
 	if abortInProgress {
-		if err := task.AbortAndMarkResetTasksForVersion(versionId, taskIds, caller); err != nil {
+		if err := task.AbortAndMarkResetTasksForVersion(versionId, abortIds, caller); err != nil {
 			return errors.WithStack(err)
 		}
 	}
-	allFinishedTasks, err := getTasksToReset(taskIds)
-	if err != nil {
-		return errors.Wrap(err, "getting finished tasks")
-	}
-	if len(allFinishedTasks) == 0 {
+	if len(tasksToReset) == 0 {
 		return nil
 	}
-	return restartTasks(ctx, allFinishedTasks, caller, versionId)
+	return restartTasks(ctx, tasksToReset, caller, versionId)
 }
 
-// getTasksToReset returns all finished tasks that should be reset given an initial input list of
+// getTasksWithDisplayTasks returns all finished tasks that should be reset given an initial input list of
 // finished taskIds. If a display task is in the list, its execution tasks are removed if they also exist.
-func getTasksToReset(taskIds []string) ([]task.Task, error) {
-	finishedTasks, err := task.FindAll(db.Query(task.ByIdsAndStatus(taskIds, evergreen.TaskCompletedStatuses)))
+func getTasksWithDisplayTasks(taskIds []string) ([]task.Task, error) {
+	tasks, err := task.FindAll(db.Query(task.ByIds(taskIds)))
 	if err != nil {
 		return nil, errors.WithStack(err)
 	}
-	allFinishedTasks, err := task.AddParentDisplayTasks(finishedTasks)
+	parents, err := task.FindAll(db.Query(task.ByExecutionTasks(taskIds)))
 	if err != nil {
-		return nil, errors.WithStack(err)
+		return nil, errors.Wrap(err, "finding parent display tasks")
 	}
-	for i := len(allFinishedTasks) - 1; i >= 0; i-- {
-		t := allFinishedTasks[i]
-		if t.DisplayTask != nil {
-			allFinishedTasks = append(allFinishedTasks[:i], allFinishedTasks[i+1:]...)
+	allTasks := []task.Task{}
+	for i, t := range tasks {
+		if !t.IsPartOfDisplay() {
```

Comment:
We may want to include exec tasks so they abort too, otherwise the parent display task will abort but won't restart until the children exec tasks finish. This is actually the current behavior when restarting a version on the UI.

Reply:
Ah good point, ty